### PR TITLE
Refactor tablet mgr code

### PR DESCRIPTION
**Pull request by yiguolei**
1. change schema hash to int32_t
2. refactor tablet mgr code to prevent dead lock
3. add tablet trash sweep thread to gc shutdown tablets
4. unify load tablet code

### CODE REVIEW

File: be/src/olap/tablet_manager.cpp [hunk 1]
```diff
@@ -872,30 +882,86 @@ OLAPStatus TabletManager::start_trash_sweep() {
...
+                            << " schema hash = " << new_tablet_meta->schema_hash();
+                // remove it from list
+                it = _shutdown_tablets.erase(it);
+                continue;
+            }
+            if (check_dir_existed((*it)->tablet_path())) {
+                // take snapshot of tablet meta
+                std::string meta_file = (*it)->tablet_path() + "/" + std::to_string((*it)->tablet_id()) + ".hdr";
+                (*it)->tablet_meta()->save(meta_file);
+                OLAPStatus rm_st = move_to_trash((*it)->tablet_path(), (*it)->tablet_path());
```
Comment (kangpinghuang) [inline be/src/olap/tablet_manager.cpp]: add INFO log before move to trash

File: be/src/olap/task/engine_storage_migration_task.cpp [hunk 2]
```diff
@@ -122,7 +123,7 @@ OLAPStatus EngineStorageMigrationTask::_storage_medium_migrate(
         // migrate all index and data files but header file
         res = _copy_index_and_data_files(schema_hash_path, tablet, consistent_rowsets);
         if (res != OLAP_SUCCESS) {
-            LOG(WARNING) << "fail to copy index and data files when migrate. res=" << res;
+            LOG(WARNING) << "fail to copy index and data files when migrate. res=" <<  res;
```
Comment (kangpinghuang) [inline be/src/olap/task/engine_storage_migration_task.cpp]: additional space

**Review by kangpinghuang — CHANGES_REQUESTED** [review]

File: be/src/olap/tablet_manager.cpp [hunk 3]
```diff
@@ -100,15 +99,22 @@ OLAPStatus TabletManager::add_tablet(TTabletId tablet_id, SchemaHash schema_hash
...
             LOG(WARNING) << "add the same tablet twice! tablet_id="
-                << tablet_id << " schema_hash=" << tablet_id;
+                         << tablet_id << " schema_hash=" << tablet_id;
+            return OLAP_ERR_ENGINE_INSERT_EXISTS_TABLE;
+        }
+    }
+
+    // if not force, the data dir should not same
+    if (!force) {
+        if (table_item->data_dir() == tablet->data_dir()) {
```
[the remark this answers is not in the document]
Comment (yiguolei) [reply be/src/olap/tablet_manager.cpp]: done

File: be/src/olap/tablet_manager.cpp [hunk 4]
```diff
@@ -81,13 +81,12 @@ TabletManager::TabletManager()
...
-OLAPStatus TabletManager::add_tablet(TTabletId tablet_id, SchemaHash schema_hash,
+OLAPStatus TabletManager::_add_tablet_unlock(TTabletId tablet_id, SchemaHash schema_hash,
                                  const TabletSharedPtr& tablet, bool force) {
     OLAPStatus res = OLAP_SUCCESS;
     VLOG(3) << "begin to add tablet to TabletManager. "
             << "tablet_id=" << tablet_id << ", schema_hash=" << schema_hash
             << ", force=" << force;
-    _tablet_map_lock.wrlock();
 
     TabletSharedPtr table_item;
```
[the remark this answers is not in the document]
Comment (yiguolei) [reply be/src/olap/tablet_manager.cpp]: done

File: be/src/olap/tablet_manager.cpp (hunk 1 above)
Comment (yiguolei) [reply be/src/olap/tablet_manager.cpp]: done

File: be/src/olap/task/engine_storage_migration_task.cpp (hunk 2 above)
Comment (yiguolei) [reply be/src/olap/task/engine_storage_migration_task.cpp]: done

**Review by kangpinghuang — APPROVED** [review]

**Review by chaoyli — CHANGES_REQUESTED** [review]

**Pull request merged**